Controllers for Request model

ar2rworld opened the pull request:
### What this PR does (required):
- Controllers for Request model manipulations
- List requests, Create request, Update request function routes
- Resolves #8

### Any information needed to test this feature   ##(required):
- You probably need to remove `protect` from
`router.route("/list").get(protect, requestList);`
`router.route("/create").post(protect, requestCreate);`
`router.route("/update").put(protect, requestUpdate);`
to test is with curl

### Any issues with the current functionality (optional):
- Should I change router to have single path ("/") and different methods to simplify?

ar2rworld commented:
_Or should I create another branch for the 2nd commit?_
### What this PR does (required):
- List, Create, Update methods created.
- Resolves #8 

### Any issues with the current functionality (optional):
- Q: Shoud I create validator for creating and updating methods?
- Q: Not sure how to handle errors thrown by methods, need some comments.

ugur-akin commented on server/controllers/request.js:
```diff
@@ -0,0 +1,58 @@
+});
+
+exports.requestUpdate = asyncHandler(async (req, res, next) => {
+  const { userId,
+    sitterId,
+    start,
+    end,
+    status
+  } = req.body
```
It isn't a bad idea to make the API more flexible but for now we can have this handler just be concerned about updating `status` (e.g. `accepted/declined`) - this will simplify things by quite a bit.

ugur-akin commented on server/controllers/request.js:
```diff
@@ -0,0 +1,58 @@
+    status
+  } = req.body
+
+  query = await Request.updateOne({
+    userId,
+    sitterId,
+    start: new Date(start),
+    end: new Date(end),
+  },{
+    status
+  }, (err, result)=>{
+    if(err){
+      res.status(404).json({ error: 'Error occurred'});
+      throw new Error("Error occurred")
+    }
+  });
```
If we provide a callback to the query, it won't return a promise (hence we can't `await` on it) - let's fix this issue.

Also, we should make `query` `const`.

ugur-akin commented on server/controllers/request.js:
```diff
@@ -0,0 +1,58 @@
+    }
+  });
+
+  res.status(200).json({ message: `request updated` });
```
It would be a good practice to return the updated object here.

ugur-akin commented on server/routes/request.js:
```diff
@@ -0,0 +1,15 @@
+} = require("../controllers/request");
+const { validateRequest } = require("../validate");
+
+router.route("/list").get(protect, requestList);
+router.route("/create").post(protect, requestCreate);
+router.route("/update").put(protect, requestUpdate);
```
Regarding your question
> Should I change router to have single path ("/") and different methods to simplify?

I think having the path as "/" for all the routes makes sense.
For `update`, assuming we apply the changes so that it only updates the status, we should change the verb (`PUT` implies we replace the whole resource). Ideally we also indicate which request we want to update by providing a [route parameter](http://expressjs.com/en/guide/routing.html#route-parameters).

ugur-akin reviewed (changes requested):
Good progress on this 🙌🏼 I highlighted some issues that needs to be fixed. I also went into a bit more detail than usual so feel free to communicate any further concerns/questions!

Also, thanks a lot for formatting your PR cleanly! The only thing I would add to it is some screenshots of the API responses, say using a tool like `postman`.

ugur-akin commented:
> Q: Shoud I create validator for creating and updating methods?

Validation is always a good idea! We can usually do most of the validation we need [within the schema](https://mongoosejs.com/docs/validation.html) , you can also implement validation logic in your controllers as necessary.

ar2rworld commented:
I think I addressed all the comments.
Some screenshots:
![image](https://user-images.githubusercontent.com/60049241/135375548-a35be6e9-089c-4158-ac11-8fab807ae506.png)
![image](https://user-images.githubusercontent.com/60049241/135375454-86146583-c468-41c0-81ee-e641a3690ab6.png)
![image](https://user-images.githubusercontent.com/60049241/135375769-df0b16e9-2b32-467b-bd4c-a01528b60dff.png)

moffatethan commented:
@ar2rworld Great job labelling your pull request! Thanks for doing that

ugur-akin commented on server/controllers/request.js:
```diff
@@ -2,57 +2,39 @@ const Request = require("../models/Request");
-      throw new Error("Error occurred")
-    }
-  });
+  const request = await Request.findOne({ _id: id });
```
This is great 🙌🏼 Alternatively, you could also use the `findById` method.

ugur-akin commented on server/controllers/request.js:
```diff
@@ -2,57 +2,39 @@ const Request = require("../models/Request");
   });
 
-  res.status(200).json({ message: 'request created'});
+  res.status(200).json({
```
Forgot to mention this before but we can use the `201` status code here instead 👍🏼

ugur-akin commented on server/routes/request.js:
```diff
@@ -6,10 +6,9 @@ const {
-router.route("/list").get(protect, requestList);
-router.route("/create").post(protect, requestCreate);
-router.route("/update").put(protect, requestUpdate);
+router.route("/").get(protect, requestList);
+router.route("/").post(protect, requestCreate);
+router.route("/:id").post(protect, requestUpdate);
```
Love these ❤️

ugur-akin reviewed (approved):
Good work on the changes, thanks 💯 🙏🏼
